### PR TITLE
feat(agent): Add background execution to the assistant drawer (hidden behind flag) (LFE-14374)

### PR DESCRIPTION
**Pull request by LysanderKie**
🟡 **Live preview: [pr-15595.preview.langfuse.com](https://pr-15595.preview.langfuse.com)**

Closes [LFE-14374](https://linear.app/clickhouse/issue/LFE-14374/drawer-cutover-to-background-execution-same-flag-stacked-on).

Builds on the background lifecycle, worker, and watch APIs merged in #15594. Extracted from the umbrella/reference PR #15558.

## Summary

Adds an opt-in background execution mode to the assistant drawer. Runs can continue in the worker when the drawer is closed or the page is refreshed, while the existing foreground path remains the default production behavior.

Internal testers enable background mode before starting a run:

```js
localStorage.setItem(
  "langfuse-in-app-agent-background-execution",
  "true",
);
location.reload();
```

Execution mode is deliberately not persisted on conversations. Do not change the local switch while a run is active.

## What changed

- Added a background AG-UI transport with persisted-cursor reconnect.
- Added `BackgroundExecutionSession` as the owner of hydration, attachment, run, cancel, approval continuation, detachment, and observable background state.
- Preserved runs across drawer close and page refresh without cancelling server execution.
- Rehydrated from persisted state before continuation attachment and after execution settles, ensuring the live transcript converges with reload state.
- Added progressive rendering for compacted blocks.
- Added current-run status, failure notices, cancellation, and persisted approval cards.
- Made Stop finish any buffered frontend reveal immediately, even when the server run has already completed.
- Kept the existing foreground `HttpAgent` path intact.

Historical inline interrupted-run notices and their placement tests are intentionally removed.

## Runtime dependency

The client switch only selects the execution path. A worker must also consume the background-run queue:

```text
QUEUE_CONSUMER_IN_APP_AGENT_RUN_QUEUE_IS_ENABLED=true
```

Without an enabled consumer, background submissions remain queued until lifecycle reconciliation marks them as timed out.

`LANGFUSE_IN_APP_AGENT_RUN_QUEUE_PROCESSING_CONCURRENCY` is optional and defaults to `5`.

Hosted-environment enablement is handled separately from this PR.

## Session guarantees

- Hydration restores the transcript and cursor from one authoritative snapshot before attachment.
- Repeated attachment is idempotent.
- Reconnect resumes above the last accepted cursor.
- Drawer close detaches observation without cancelling execution.
- Cancellation targets the active server run.
- Approval decisions rehydrate before continuation attachment.
- Terminal execution rehydrates from persisted state so live rendering matches reload.
- Watch failure remains distinct from server-run completion.
- Repeated no-progress watches back off and terminate instead of hot-looping.

## Deliberate follow-ups

Before enabling customers:

- allow conversation switching by detaching observation without cancellation;
- complete the unclaimed-run redispatch guarantee;
- enforce the organization active-run ceiling;
- validate worker capacity and 100 concurrent watches;
- finish credential-expiry and operational-cost validation.

## Impact

- `web`: background transport/session integration and drawer rendering.
- No persisted execution-mode field and no change to the default foreground path.

## Verification

- Focused in-app-agent client suite: `Test Files 9 passed (9)`, `Tests 88 passed (88)`.
- Background session suite: `Test Files 1 passed (1)`, `Tests 14 passed (14)`.
- Local background-mode browser flow:
  - two sequential approval-gated actions completed;
  - all tool calls remained visible before reload;
  - the expanded live and reloaded transcripts were exactly equal;
  - no visible drawer regression was observed.
- Latest CI head: lint, formatting, web/worker builds, shared tests, worker tests, Storybook, E2E, CodeQL, and security checks passed.

## Tests added: 14

`web/src/features/in-app-agent/lib/backgroundExecutionSession.clienttest.ts` covers hydration and attachment ordering, detach without cancellation, cancellation, approval continuations, attachment failures, immediate queued state, cursor reconnect, no-progress backoff, invalid watch frames, and live transcript convergence across sequential approvals.

Three smooth-streaming tests were updated to cover compacted-block rendering, hydrated approvals, and Stop finishing a buffered response immediately.

<details open><summary><h3>Greptile Summary</h3></summary>

Adds an opt-in worker-backed execution path for the in-app assistant.
- Introduces a persisted-cursor AG-UI client and session controller for hydration, attachment, cancellation, approvals, and transcript convergence.
- Integrates background lifecycle state and Stop controls into the assistant drawer while retaining the foreground transport.
- Updates smooth-stream rendering for compacted blocks and adds focused background-session tests.
- Adds RxJS as a direct web dependency.
</details>

<details><summary><h3>Confidence Score: 4/5</h3></summary>

The active-run lifecycle notice must be moved outside the turn-progress-hidden container before merging so queued, running, and stopping guidance is actually visible.

The background execution lifecycle is broadly integrated, but every notice intended for an active run is collapsed and marked aria-hidden under exactly those active states, defeating a user-facing part of the feature.

**Files Needing Attention:** web/src/features/in-app-agent/components/InAppAgentWindow.tsx
</details>

<details><summary><h3>Sequence Diagram</h3></summary>

```mermaid
sequenceDiagram
  participant UI as Assistant drawer
  participant Session as Background session
  participant API as tRPC / watch API
  participant Worker as Background worker
  UI->>Session: Submit message
  Session->>API: Start run
  API->>Worker: Queue run
  Session->>API: Watch from persisted cursor
  Worker-->>API: Persist status and AG-UI events
  API-->>Session: Status and event frames
  Session-->>UI: Update transcript and lifecycle state
  UI->>Session: Detach on drawer close
  Note over Session,Worker: Server execution continues
  UI->>Session: Rehydrate and reattach
  Session->>API: Load snapshot, then watch above cursor
```
</details>

<details><summary>Prompt To Fix All With AI</summary>

`````markdown
### Issue 1
web/src/features/in-app-agent/components/InAppAgentWindow.tsx:874-885
**Active-run notices stay hidden**

When a background run is queued, running, or stopping, `backgroundRunNotice` is rendered inside the container that receives `aria-hidden`, `max-h-0`, and `opacity-0` for the entire assistant turn, so users never see the guidance that execution continues after closing the drawer.

---

For each issue above, determine whether it is valid and should be fixed. If so, fix it directly.
`````

</details>

<sub>Reviews (1): Last reviewed commit: ["Merge branch &#39;main&#39; into feature/lfe-143..."](https://github.com/langfuse/langfuse/commit/a4d227c46cc8322306c97bea7752f3f161d137ac) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=48689747)</sub>

> Greptile also left **1 inline comment** on this PR.

### CODE REVIEW

File: web/src/features/in-app-agent/components/InAppAgentWindow.tsx [hunk 1]
```diff
@@ -854,6 +871,19 @@ export function InAppAgentWindow(props: InAppAgentWindowProps) {
...
+                  className={cn(
+                    "border-border bg-muted/60 text-foreground-secondary flex w-full items-center gap-1 rounded-lg border px-2 py-1",
+                    isExpanded ? "text-sm" : "text-xs",
+                  )}
+                >
+                  <Info aria-hidden="true" className="size-3 shrink-0" />
+                  <span className="min-w-0" title={backgroundRunNotice}>
+                    {backgroundRunNotice}
+                  </span>
+                </p>
```
[the remark this answers is not in the document]
Comment (LysanderKie) [reply web/src/features/in-app-agent/components/InAppAgentWindow.tsx]: Fixed in d7614f521. The background lifecycle notice now renders outside the active-turn-collapsed wrapper, uses `role="status"`, and uses the valid `text-foreground-tertiary` token. The window behavior test and seeded browser review both verify the notice and Stop remain visible during an active background run.

File: web/src/features/in-app-agent/components/InAppAgentWindow.tsx [hunk 2]
```diff
@@ -854,6 +871,19 @@ export function InAppAgentWindow(props: InAppAgentWindowProps) {
...
+              )}
```
[the remark this answers is not in the document]
Comment (LysanderKie) [reply web/src/features/in-app-agent/components/InAppAgentWindow.tsx]: Fixed in d7614f521. The background lifecycle notice now renders outside the active-turn-collapsed wrapper, uses `role="status"`, and uses the valid `text-foreground-tertiary` token. The window behavior test and seeded browser review both verify the notice and Stop remain visible during an active background run.

File: web/src/features/in-app-agent/lib/backgroundExecutionSession.ts [hunk 3]
```diff
@@ -0,0 +1,495 @@
...
+      currentRun: { ...run, cancelRequested: true },
+    });
+    await this.hydrateAndAttach();
+  }
+
+  async decide(input: ApprovalDecision): Promise<void> {
+    await this.decideApproval(input);
+    this.detach();
+    await this.hydrateAndAttach();
+  }
```
[the remark this answers is not in the document]
Comment (LysanderKie) [reply web/src/features/in-app-agent/lib/backgroundExecutionSession.ts:284]: Fixed in d7614f521. Approval and cancellation promises now resolve according to the durable mutation. The session removes accepted approvals / records accepted cancellation immediately, then treats rehydration as best-effort attachment work; refresh failures remain in the session snapshot and cannot resurrect the action or produce a false mutation failure. Focused regressions cover both accepted-command/failed-refresh cases and mutation-failure restoration.

**Comment by LysanderKie** [comment]
## Hardening pass complete

Implemented in `d7614f521`:

- Background messages, approvals, current run, attachment, and cancellation UI state now flow through one coherent external-store snapshot; the foreground `HttpAgent` path remains unchanged and default-off behavior is covered.
- Approval/cancel commands now resolve at the durable mutation boundary. Accepted commands survive a failed refresh; mutation failures restore actionable UI.
- The drawer receives a discriminated foreground/background execution UI. Active lifecycle copy and Stop remain visible during generation, and Stop disables while submitting.
- Expected background transport/hydration/run failures no longer emit `console.error`; user-visible recovery state remains.
- README now documents both execution paths, ownership, detach semantics, rollout/sunset, and the explicit decision to add **no Stop analytics event** during the internal rollout.

Rollout prerequisites split from LFE-14374 and linked as blockers of LFE-14375:

- [LFE-14642 — Retry with prefilled input](https://linear.app/clickhouse/issue/LFE-14642/background-run-retry-with-prefilled-input)
- [LFE-14643 — Conversation run-status badges and scoped polling](https://linear.app/clickhouse/issue/LFE-14643/conversation-background-run-status-badges-and-scoped-polling)

Verification:

- Focused client suites: `Test Files 3 passed (3)`, `Tests 25 passed (25)`.
- Repository lint: `Tasks: 7 successful, 7 total`.
- Repository typecheck: `Tasks: 7 successful, 7 total`.
- Seeded real-browser review covered foreground submit/response/close-reopen and background queued/running notice, detach/reattach without duplicate transcript entries, refresh persistence, approval rejection/continuation, cancellation, and terminal transcript. Browser console errors: 0.

Independent Standards and Spec re-reviews found no remaining actionable blockers.

**Comment by LysanderKie** [comment]
CI follow-up: the standard web matrix was retried and reproduced the sole failure in `LangfuseMediaView.clienttest.tsx`; all 2,821 other client tests passed, including the 25 in-app-agent tests. The exact `main` base commit (`2e089efa`) fails on the same test in [run 30557182733](https://github.com/langfuse/langfuse/actions/runs/30557182733/job/90920621029), so this is a base-branch failure rather than a regression from this PR. No unrelated media fix has been pulled into this already-large PR.

**Comment by LysanderKie** [comment]
@claude review

File: web/src/features/in-app-agent/components/InAppAiAgentProvider.tsx [hunk 4]
```diff
@@ -721,34 +875,187 @@ function InAppAiAgentProviderInner({
...
+                    id: snapshot.latestRun.id,
+                    status: snapshot.latestRun.status,
+                    errorCode: snapshot.latestRun.errorCode,
+                    cancelRequested: snapshot.latestRun.cancelRequested,
+                  }
+                : null,
+              pendingToolApprovals: snapshot.pendingToolApprovals.map(
+                (approval) => ({ ...approval, status: "pending" as const }),
+              ),
+            } satisfies Omit<
```
[the remark this answers is not in the document]
Comment (LysanderKie) [reply web/src/features/in-app-agent/components/InAppAiAgentProvider.tsx:969]: Valid finding, but fixed differently than proposed, in #15661.

`getConversationMessagesForReplay` is not the right accessor here: it strips reasoning, redirect action results, run IDs and unpaired tool calls because it exists to build model context. Swapping to it would replace one representation leak with another.

The actual defect is that the display projection was folded twice. It truncates an assistant message at its first interleaved block and moves the continuation into a synthetic `display-text-<id>-N` sibling, which is right for rendering and wrong for anything else. The server folded it, then the client seeded its live AG-UI agent with the result.

The client already folds canonical to display for live messages, so the fix is to stop folding on the server: `getConversation` now returns canonical messages plus the display state as a serialized sidecar, and the projection runs exactly once, at render time. Display recording and projection also moved out of `packages/shared` into the web feature folder, since the worker never renders.

Regression test drives the real `InAppAgentBackgroundClient` over a mocked SSE tail so the actual AG-UI apply pipeline runs: hydrate `text -> thought -> continued text`, deliver a further delta for the original assistant message, assert the canonical message stays whole and the projection still renders initial text, thought, complete continuation. Confirmed red against the old contract first, failing with exactly the reported symptom (`"Initial answer. And one more."`).

File: web/src/features/in-app-agent/lib/backgroundExecutionSession.ts [hunk 5]
```diff
@@ -0,0 +1,545 @@
...
+      throw error;
+    }
+    this.setView({
+      ...this.view,
+      currentRun: { ...run, cancelRequested: true },
+      cancelStatus: "idle",
+    });
+    this.detach();
+    await this.refreshAttachmentAfterCommand();
+  }
```
[the remark this answers is not in the document]
Comment (LysanderKie) [reply web/src/features/in-app-agent/lib/backgroundExecutionSession.ts:271]: Confirmed and fixed in #15661.

`cancel()` now re-reads `this.view.currentRun` after the mutation resolves and stamps `cancelRequested` only when the id still matches, following the read-modify-write-at-call-time shape `decide()` already used. A superseding run is left untouched.

Two regression tests: one delivers a terminal status through the watch listener while the cancel mutation is in flight and asserts the view never regresses to RUNNING; the other covers the matching-id branch. Both needed a deferred hydrate, because the existing cancel test's hydrate mock returned an already-cancelled view, which is why it masked this.

**Comment by LysanderKie** [comment]
close as https://github.com/langfuse/langfuse/pull/15661 merged as replacement
